Content.Server/Execution/ExecutionSystem.cs

```diff
@@ -0,0 +1,241 @@
+    [Dependency] private readonly SharedMeleeWeaponSystem _meleeSystem = default!;
+
+    private const float MeleeExecutionTimeModifier = 5.0f;
+    private const float GunExecutionTime = 10.0f;
+    private const float DamageModifier = 9.0f;
+
+    /// <inheritdoc/>
+    public override void Initialize()
+    {
+        base.Initialize();
+        
+        SubscribeLocalEvent<SharpComponent, GetVerbsEvent<UtilityVerb>>(OnGetInteractionVerbsMelee);
+        SubscribeLocalEvent<GunComponent, GetVerbsEvent<UtilityVerb>>(OnGetInteractionVerbsGun);
+        
+        SubscribeLocalEvent<SharpComponent, ExecutionDoAfterEvent>(OnDoafterMelee);
+        SubscribeLocalEvent<GunComponent, ExecutionDoAfterEvent>(OnDoafterGun);
+    }
+
+    private void OnGetInteractionVerbsMelee(
+        EntityUid uid, 
+        SharpComponent component,
+        GetVerbsEvent<UtilityVerb> args)
+    {
+        if (args.Hands == null || args.Using == null || !args.CanAccess || !args.CanInteract)
+            return;
+        
+        var attacker = args.User;
+        var weapon = args.Using!.Value;
+        var victim = args.Target;
+
+        if (!CanExecuteWithMelee(weapon, victim, victim))
+            return;
+        
+        UtilityVerb verb = new()
+        {
+            Act = () =>
+            {
+                TryStartMeleeExecutionDoafter(weapon, victim, attacker);
+            },
+            Impact = LogImpact.High,
+            Text = Loc.GetString("execution-verb-name"),
+            Message = Loc.GetString("execution-verb-message"),
+        };
+
+        args.Verbs.Add(verb);
+    }
+
+    private void OnGetInteractionVerbsGun(
+        EntityUid uid, 
+        GunComponent component,
+        GetVerbsEvent<UtilityVerb> args)
+    {
+        if (args.Hands == null || args.Using == null || !args.CanAccess || !args.CanInteract)
+            return;
+
+        var attacker = args.User;
+        var weapon = args.Using!.Value;
+        var victim = args.Target;
+
+        if (!CanExecuteWithGun(weapon, victim, victim))
+            return;
+        
+        UtilityVerb verb = new()
+        {
+            Act = () =>
+            {
+                TryStartGunExecutionDoafter(weapon, victim, attacker);
+            },
+            Impact = LogImpact.High,
+            Text = Loc.GetString("execution-verb-name"),
+            Message = Loc.GetString("execution-verb-message"),
+        };
+
+        args.Verbs.Add(verb);
+    }
+
+    private bool CanExecuteWithAny(EntityUid weapon, EntityUid victim, EntityUid user)
+    {
+        // No point executing someone if they can't take damage
+        if (!TryComp<DamageableComponent>(victim, out var damage))
+            return false;
+        
+        // You're not allowed to execute dead people (no fun allowed)
+        if (TryComp<MobStateComponent>(victim, out var mobState) && _mobStateSystem.IsDead(victim, mobState))
+            return false;
```

Comment:
it should probably additionally require mobstate on the target like you shouldnt be able to execute walls

Reply:
yes, the current code does not require mobstate

Reply:
can you provide a code suggestion because i don't see it

Reply:
i tested it, you definitely can't execute a wall

Reply:
well that probably just fails separately because of the caninteract check but itd be preferable to actually check for mobstate